Bad error message on alloc exec when streaming through ELB

### Nomad version
Nomad v0.10.0-rc1 (c49bf41779d2ee462f46e2f58328e58c51970a83)

### Operating system and Environment details
Linux in AWS EC2

### Issue
Nomad alloc exec does not work through EC2 Classic ELB. However, the error message returned in this scenario is not human-readable, complicating the ability for operators to diagnose this issue.

### Reproduction steps
1. Spin up Nomad cluster in EC2 (for example, using TF config in `/terraform`)
2. Front the server(s) with a classic ELB, with a listener configured for `HTTP/4646` (Layer 7)
3. Note that `alloc fs` commands work.
4. Note that `alloc exec` commands do not: 
```
$ nomad alloc exec -job nginx  bash
�0iled to exec into task: Unexpected response code: 400 (�<��
�mHG���������N�x,E��ڋ��XC�2~�
G���-#]���v�O�!��H��"����H�#U�)##6�6�t�<0��z��
```

### Nomad Server log

When encountering the error, the server emits the following error message: 
```
Oct 24 15:43:40 ip-172-31-41-217 nomad[2198]:     2019-10-24T15:43:40.285Z [ERROR] http: request failed: method=GET path=/v1/client/allocation/ea145ac4-56e7-7c45-47a1-2ad3dfb334d2/exec?command=%5B%22bash%22%5D&task=nginx&tty=true error="failed to upgrade connection: websocket: the client is not using the websocket protocol: 'upgrade' token not found in 'Connection' header" code=500
```

### Summary

The root cause is that Level 7 (HTTP) ELBs in EC2 do not support WebSockets. Readers should note that there are multiple approaches to resolving this: change the listener to Level 4 (TCP); or use the newer EC2 Application Load Balancer (ALB). Regardless, the error message emitted by the CLI should propagate the error from the server, if possible.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ BUG FIXES:
  * nomad: Multiple connect enabled services in the same taskgroup failed to
    register [[GH-6646](https://github.com/hashicorp/nomad/issues/6646)]
  * scheduler: Changes to devices in resource stanza should cause rescheduling [[GH-6644](https://github.com/hashicorp/nomad/issues/6644)]
+ * api: Decompress web socket response body if gzipped on error responses [[GH-6650](https://github.com/hashicorp/nomad/issues/6650)]
  
 ## 0.10.1 (November 4, 2019)
 

--- a/api/allocations_test.go
+++ b/api/allocations_test.go
@@ -9,7 +9,7 @@ import (
 
 	"time"
 
-	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/nomad/helper/uuid"
 	"github.com/stretchr/testify/require"
 )
 
@@ -148,20 +148,12 @@ func TestAllocations_RescheduleInfo(t *testing.T) {
 	}
 	job.Canonicalize()
 
-	uuidGen := func() string {
-		ret, err := uuid.GenerateUUID()
-		if err != nil {
-			t.Fatal(err)
-		}
-		return ret
-	}
-
 	alloc := &Allocation{
-		ID:        uuidGen(),
+		ID:        uuid.Generate(),
 		Namespace: DefaultNamespace,
-		EvalID:    uuidGen(),
+		EvalID:    uuid.Generate(),
 		Name:      "foo-bar[1]",
-		NodeID:    uuidGen(),
+		NodeID:    uuid.Generate(),
 		TaskGroup: *job.TaskGroups[0].Name,
 		JobID:     *job.ID,
 		Job:       job,
@@ -273,20 +265,12 @@ func TestAllocations_ExecErrors(t *testing.T) {
 	}
 	job.Canonicalize()
 
-	uuidGen := func() string {
-		ret, err := uuid.GenerateUUID()
-		if err != nil {
-			t.Fatal(err)
-		}
-		return ret
-	}
-
 	alloc := &Allocation{
 		ID:        "",
 		Namespace: DefaultNamespace,
-		EvalID:    uuidGen(),
+		EvalID:    uuid.Generate(),
 		Name:      "foo-bar[1]",
-		NodeID:    uuidGen(),
+		NodeID:    uuid.Generate(),
 		TaskGroup: *job.TaskGroups[0].Name,
 		JobID:     *job.ID,
 		Job:       job,

--- a/api/allocations_test.go
+++ b/api/allocations_test.go
@@ -1,6 +1,8 @@
 package api
 
 import (
+	"context"
+	"os"
 	"reflect"
 	"sort"
 	"testing"
@@ -245,6 +247,58 @@ func TestAllocations_RescheduleInfo(t *testing.T) {
 		})
 	}
 
+}
+
+// TestAllocations_ExecErrors ensures errors are properly formatted
+func TestAllocations_ExecErrors(t *testing.T) {
+	c, s := makeClient(t, nil, nil)
+	defer s.Stop()
+	a := c.Allocations()
+
+	job := &Job{
+		Name:      stringToPtr("foo"),
+		Namespace: stringToPtr(DefaultNamespace),
+		ID:        stringToPtr("bar"),
+		ParentID:  stringToPtr("lol"),
+		TaskGroups: []*TaskGroup{
+			{
+				Name: stringToPtr("bar"),
+				Tasks: []*Task{
+					{
+						Name: "task1",
+					},
+				},
+			},
+		},
+	}
+	job.Canonicalize()
+
+	uuidGen := func() string {
+		ret, err := uuid.GenerateUUID()
+		if err != nil {
+			t.Fatal(err)
+		}
+		return ret
+	}
+
+	alloc := &Allocation{
+		ID:        "",
+		Namespace: DefaultNamespace,
+		EvalID:    uuidGen(),
+		Name:      "foo-bar[1]",
+		NodeID:    uuidGen(),
+		TaskGroup: *job.TaskGroups[0].Name,
+		JobID:     *job.ID,
+		Job:       job,
+	}
+	// Querying when no allocs exist returns nothing
+	sizeCh := make(chan TerminalSize, 1)
+
+	// make a request that will result in an error
+	// ensure the error is what we expect
+	_, err := a.Exec(context.Background(), alloc, "bar", false, []string{"command"}, os.Stdin, os.Stdout, os.Stderr, sizeCh, nil)
+	require.Contains(t, err.Error(), "Unexpected response code: 301")
+	require.Contains(t, err.Error(), "Moved Permanently")
 }
 
 func TestAllocations_ShouldMigrate(t *testing.T) {

--- a/api/api.go
+++ b/api/api.go
@@ -742,7 +742,16 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
 	// check resp status code, as it's more informative than handshake error we get from ws library
 	if resp != nil && resp.StatusCode != 101 {
 		var buf bytes.Buffer
-		io.Copy(&buf, resp.Body)
+
+		if resp.Header.Get("Content-Encoding") == "gzip" {
+			greader, err := gzip.NewReader(resp.Body)
+			if err != nil {
+				return nil, nil, fmt.Errorf("Unexpected response code: %d", resp.StatusCode)
+			}
+			io.Copy(&buf, greader)
+		} else {
+			io.Copy(&buf, resp.Body)
+		}
 		resp.Body.Close()
 
 		return nil, nil, fmt.Errorf("Unexpected response code: %d (%s)", resp.StatusCode, buf.Bytes())

